raft/src/test/java/org/apache/kafka/raft/InvalidMemoryRecordsProvider.java

```diff
@@ -0,0 +1,143 @@
+    public static final int EPOCH = 4321;
+
+    // TODO: use jqwik support for random generators
+    public static MemoryRecords buildRandomRecords(Random random) {
+        int size = random.nextInt(255) + 1;
+        byte[] bytes = new byte[size];
+        random.nextBytes(bytes);
+
+        return MemoryRecords.readableRecords(ByteBuffer.wrap(bytes));
+    }
+
+    /** Returns a stream of arguements for invalid memory records and the expected exception.
+     *
+     * The first object in the Arguments is a MemoryRecords.
+     *
+     * The second object in the Arguments is an Class<Exception> which is the expected exception from the log layer
+     */
+    @Override
+    public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
+        return Stream.of(
+            Arguments.of(MemoryRecords.readableRecords(notEnoughtBytes()), CorruptRecordException.class),
+            Arguments.of(MemoryRecords.readableRecords(recordsSizeTooSmall()), CorruptRecordException.class),
+            Arguments.of(MemoryRecords.readableRecords(notEnoughBytesToMagic()), CorruptRecordException.class),
+            Arguments.of(MemoryRecords.readableRecords(negativeMagic()), CorruptRecordException.class),
+            Arguments.of(MemoryRecords.readableRecords(largeMagic()), CorruptRecordException.class),
+            Arguments.of(MemoryRecords.readableRecords(lessBytesThanRecordSize()), CorruptRecordException.class)
+        );
+    }
+
+    private static ByteBuffer notEnoughtBytes() {
+        var buffer = ByteBuffer.allocate(Records.LOG_OVERHEAD - 1);
+        buffer.limit(buffer.capacity());
+
+        return buffer;
+    }
+
+    private static ByteBuffer recordsSizeTooSmall() {
+        var buffer = ByteBuffer.allocate(256);
+        // Write the base offset
+        buffer.putLong(BASE_OFFSET);
+        // Write record size
+        buffer.putInt(LegacyRecord.RECORD_OVERHEAD_V0 - 1);
+        buffer.position(0);
+        buffer.limit(buffer.capacity());
+
+        return buffer;
+    }
+
+    private static ByteBuffer notEnoughBytesToMagic() {
+        var buffer = ByteBuffer.allocate(256);
+        // Write the base offset
+        buffer.putLong(BASE_OFFSET);
+        // Write record size
+        buffer.putInt(buffer.capacity() - Records.LOG_OVERHEAD);
+        buffer.position(0);
+        buffer.limit(Records.HEADER_SIZE_UP_TO_MAGIC - 1);
+
+        return buffer;
+    }
+
+    private static ByteBuffer negativeMagic() {
+        var buffer = ByteBuffer.allocate(256);
+        // Write the base offset
+        buffer.putLong(BASE_OFFSET);
+        // Write record size
+        buffer.putInt(buffer.capacity() - Records.LOG_OVERHEAD);
+        // Write the epoch
+        buffer.putInt(EPOCH);
+        // Write magic
+        buffer.put((byte) -1);
+        buffer.position(0);
+        buffer.limit(buffer.capacity());
+
+        return buffer;
+    }
+
+    private static ByteBuffer largeMagic() {
```

Comment:
nit: should this just be incorrectMagic

Reply:
Large magic is one incorrect magic. Another incorrect magic is a negative number. Hence why I added `negativeMagic` and `largeMagic`.